scripts/release.sh

```diff
@@ -0,0 +1,201 @@
+#
+# Preprequisites:
+# Before running this script make sure "cargo-release" software in installed:
+# https://github.com/crate-ci/cargo-release
+# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+line="- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -"
+
+WORK_STATUS=$(git status --porcelain)
+if [ "${WORK_STATUS}" != "" ];
+then
+  echo $line
+  echo "Working directory must be clean. Currently it contains some changes!"
+  echo $WORK_STATUS
+  echo $line
+  exit 1
+fi
+
+# Are we allowed to execute this script on the branch other than master?
+ALLOW_BRANCH="n"
+for param in $1
+do
+  case "$param" in
+    "--allow-branch")
+      ALLOW_BRANCH="yes"
+      echo "Execution of this script on non-master branch is allowed"
+    ;;
+  esac
+done
+CURRENT_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+
+if [ "$ALLOW_BRANCH" == "n" ];
+then
+
+  if [ "${CURRENT_BRANCH}" != "master" ];
+  then
+    echo $line
+    echo "Current branch must be master!"
+    echo $CURRENT_BRANCH
+    echo $line
+    exit 1
+  else
+    # refresh master branch
+    echo "Fetching master from remote..."
+    git fetch
+    git pull
+
+    returnStatus=$(($?+0))
+    if [ $returnStatus -ne 0 ];
+    then
+      exit $returnStatus
+    fi
+
+    echo "Comparing current master with remote..."
+    PUSH_STATUS=$(git diff HEAD origin/master --name-status)
+    if [ "${PUSH_STATUS}" != "" ];
+    then
+      echo "There are comitted but not pushed changes in your current branch!!!"
+      exit 1
+    fi
+  fi  # end of when on master
+
+fi # end of allow branch
+
+LATEST_TAG=$(git describe --abbrev=0 --tags)
+LATEST_SEMVER=${LATEST_TAG/v/}
+
+echo $line
+echo ""
+echo "CURRENT_BRANCH=$CURRENT_BRANCH"
+echo "LATEST_TAG=$LATEST_TAG"
+echo "LATEST_SEMVER=$LATEST_SEMVER"
+echo ""
+
+echo "Provide a new semver number. Type: 'a' or just press 'enter' for cancel and abort here."
+echo "Otherwise type version number, for example, type '2.0.0' the tag name will be v2.0.0"
+unset ANSWER
+read ANSWER
+if [ "${ANSWER}" == "a" ] || [ "${ANSWER}" == "" ];
+then
+  echo "Your answer was '${ANSWER}', aborting"
+  exit 1
+fi
+
+echo "You answered '$ANSWER', validating ..."
+echo ""
+
+IFS='.' read -a segments <<< "$ANSWER"
+len=${#segments[@]}
+if [ $len -ne 3 ];
+then
+  echo "Invalid version format: '${ANSWER}'. We expect exactly three segments separated by dot."
+  exit 1
+fi
+
+NEW_VERSION=$ANSWER
+
+echo $line
+echo "Provide the next version number (aka 'dev version' or 'shapshot')"
+echo "Type: 'a' or just press 'enter' for cancel and abort here."
+echo "Otherwise type the next version (witout suffix), for example, if you are releasing '2.0.0' then the next dev version will be '2.1.0'. We will add '-dev' suffix automatically"
+unset ANSWER
+read ANSWER
+if [ "${ANSWER}" == "a" ] || [ "${ANSWER}" == "" ];
+then
+  echo "Your answer was '${ANSWER}', aborting"
+  exit 1
+fi
+
+NEXT_VERSION=$(echo $ANSWER | sed 's/-dev//')-dev
```

Comment:
We are not tagging dev version.
Why we are creating was explained on Thu in our meeting. In short, the previous version was already tagged, we everything what happens with sources (like development and testing of additional commits) after that does not belong to tagged version any more. Its a standard process.

Reply:
This is not really a standard practice I have seen, that is why I would like to understand more on this. 

Looking at the script, between the tagging of `NEW_VERSION` and `NEXT_VERSION` the only commit inserted is this one here:-

```bash
git commit -a -m "chore(npm): Release npm $NEW_VERSION" --no-verify
```

There is really no need to have `NEXT_VERSION`, We can always look at the last `tag` and interpret everything from then is not published. 

You can look at `tokio`, `serde` or other popular packages, not just in rust but js/other languages as well, there are subsequent commits in `master` since the last release, but the version is not changed to some `alpha, beta, dev..etc`. 

If something version needs to be tested as `alpha, beta..etc`, that is only when we create those tags.

Reply:
> Looking at the script, between the tagging of NEW_VERSION and NEXT_VERSION the only commit inserted is this one here

There is also `cargo release` line which invokes cargo-release tool. This tool is changing versions in all `Cargo.toml` files and committing that change. Obviously it cannot do it for NPM modules, so we do it ourselves using the line you copied in your comment.

Reply:
> There is really no need to have NEXT_VERSION

Properly written apps include version number into bundled build. For example, a website may have an invisible watermark in the footer section saying current version is vXYZ, or sent into Google Analytics and etc.  The API server may include http header with x-app-version: vXYZ. This info is also used when tagging log lines which are sent to observability aggregator. This information is taken from metadata included in the executable. That metadata typically comes from package.json or from Cargo.toml or from build.gradle (in Java) and etc. When we are running the app which is built after TAG we want to show the true version. If package.json was not updated then app would display NEW_VERSION, but it may be incorrect if between TAG and now there are commits. 

I don't understand why in the meeting you had no problems with it, but after I spent time implementing now it seems you are agains a simple version bump with -dev suffix? What harm does it do?

Reply:
I probably didn't understand the usage of the `NEXT_VERSION` well, till I actually looked into the implementation. 

I agree on having the `dev` or `beta` etc tag is helpful, but I just feel we shouldn't make it strict/mandatory to have the second tag, on the release commit, but only have it on need basis. I think this is the part I must have not understood in the last meeting.